pkg/logql/log/labels_test.go

```diff
@@ -46,3 +48,102 @@ func TestLabelsBuilder_LabelsError(t *testing.T) {
 	// make sure the original labels is unchanged.
 	require.Equal(t, labels.Labels{labels.Label{Name: "already", Value: "in"}}, lbs)
 }
+
+func TestLabelsBuilder_LabelsResult(t *testing.T) {
+	lbs := labels.Labels{
+		labels.Label{Name: "namespace", Value: "loki"},
+		labels.Label{Name: "job", Value: "us-central1/loki"},
+		labels.Label{Name: "cluster", Value: "us-central1"},
+	}
+	sort.Sort(lbs)
+	b := NewBaseLabelsBuilder().ForLabels(lbs, lbs.Hash())
+	b.Reset()
+	assertLabelResult(t, lbs, b.LabelsResult())
+	b.SetErr("err")
+	withErr := append(lbs, labels.Label{Name: ErrorLabel, Value: "err"})
+	sort.Sort(withErr)
+	assertLabelResult(t, withErr, b.LabelsResult())
+
+	b.Set("foo", "bar")
+	b.Set("namespace", "tempo")
+	b.Set("buzz", "fuzz")
+	b.Del("job")
+	expected := labels.Labels{
+		labels.Label{Name: ErrorLabel, Value: "err"},
+		labels.Label{Name: "namespace", Value: "tempo"},
+		labels.Label{Name: "cluster", Value: "us-central1"},
+		labels.Label{Name: "foo", Value: "bar"},
+		labels.Label{Name: "buzz", Value: "fuzz"},
+	}
+	sort.Sort(expected)
+	assertLabelResult(t, expected, b.LabelsResult())
+	// cached.
+	assertLabelResult(t, expected, b.LabelsResult())
+}
+
+func TestLabelsBuilder_GroupedLabelsResult(t *testing.T) {
+	lbs := labels.Labels{
+		labels.Label{Name: "namespace", Value: "loki"},
+		labels.Label{Name: "job", Value: "us-central1/loki"},
+		labels.Label{Name: "cluster", Value: "us-central1"},
+	}
+	sort.Sort(lbs)
+	b := NewBaseLabelsBuilderWithGrouping([]string{"namespace"}, false, false).ForLabels(lbs, lbs.Hash())
+	b.Reset()
+	assertLabelResult(t, labels.Labels{labels.Label{Name: "namespace", Value: "loki"}}, b.GroupedLabels())
+	b.SetErr("err")
+	withErr := append(lbs, labels.Label{Name: ErrorLabel, Value: "err"})
+	sort.Sort(withErr)
+	assertLabelResult(t, withErr, b.GroupedLabels())
+
+	b.Reset()
+	b.Set("foo", "bar")
+	b.Set("namespace", "tempo")
+	b.Set("buzz", "fuzz")
+	b.Del("job")
+	expected := labels.Labels{
+		labels.Label{Name: "namespace", Value: "tempo"},
+	}
+	sort.Sort(expected)
+	assertLabelResult(t, expected, b.GroupedLabels())
+	// cached.
+	assertLabelResult(t, expected, b.GroupedLabels())
+
+	b = NewBaseLabelsBuilderWithGrouping([]string{"job"}, false, false).ForLabels(lbs, lbs.Hash())
+	assertLabelResult(t, labels.Labels{labels.Label{Name: "job", Value: "us-central1/loki"}}, b.GroupedLabels())
+	assertLabelResult(t, labels.Labels{labels.Label{Name: "job", Value: "us-central1/loki"}}, b.GroupedLabels())
```

Comment:
Is it required to check this twice or we can remove one?

Reply:
Just checking the cache is working correctly on the second pass so yes.